doc/oidentd.8
.\" oidentd.8 - oidentd ident (RFC 1413) implementation.
.\" Copyright (c) 2001-2006 Ryan McCabe <ryan@numb.org>
.\" Copyright (c) 2018      Janik Rabe  <oidentd@janikrabe.com>
.\"
.\" Permission is granted to copy, distribute and/or modify this document
.\" under the terms of the GNU Free Documentation License, Version 1.1
.\" or any later version published by the Free Software Foundation;
.\" with the Invariant Sections being no invariant sections, with the
.\" Front-Cover Texts being no front-cover texts, and with the Back-Cover
.\" Texts being no back-cover texts. A copy of the license is included with
.\" this package in the file "COPYING.DOC."
.TH oidentd 8 "2019-01-10" "version 2.3.2"

.SH NAME
oidentd \- TCP/IP Ident protocol server

.SH SYNOPSIS
.BI "oidentd [options]"
.sp
.B [ \-dehiImoqSv ]
.br
.B [ \-a <host> ]
.br
.B [ \-c <charset> ]
.br
.B [ \-C <config file> ]
.br
.B [ \-f <port> ]
.br
.B [ \-p <port> ]
.br
.B [ \-P <host> ]
.br
.B [ \-o or \-\-other=[<OS string>] ]
.br
.B [ \-t or \-\-timeout=<seconds> ]
.br
.B [ \-g or \-\-group=<group|GID> ]
.br
.B [ \-l or \-\-limit=<number>]
.br
.B [ \-r or \-\-reply=<string> ]
.br
.B [ \-u or \-\-user=<username|UID> ]

.SH DESCRIPTION
\fBoidentd\fP is a server that implements the Identification Protocol as
specified in RFC 1413.
.PP
\fBoidentd\fP operates by looking up specific TCP connections and returning
the user name of the process owning the connection.

.SH OPTIONS

.TP
.B "\-a or \-\-address=<address|hostname>"
Listen for connections on the specified address. If this option is specified
multiple times, \fBoidentd\fP will listen on all specified IP addresses. The
default is to listen for connections on all configured IP addresses.

.TP
.B "\-c or \-\-charset=<charset>"
Use the specified alternate charset.

.TP
.B "\-C or \-\-config=<config file>"
Use the specified file as the configuration file. The default location of the
configuration file is \fB/etc/oidentd.conf\fP.

.TP
.B "\-d or \-\-debug"
Enable debugging. This causes debugging messages to be printed via syslog. This
option can be useful when trying to track down the cause of failed lookups.
Note that this option is only available if \fBoidentd\fP has been compiled with
the \fB--enable-debug\fP flag.

.TP
.B "\-e or \-\-error"
Return "UNKNOWN-ERROR" for all errors, so as not to divulge any unnecessary
information to remote clients.

.TP
.B "\-f or \-\-forward=[<port>]"
When IP masquerading support is enabled, forward requests for machines that
masquerade through us to those machines on the specified port. If a port is not
given, \fBoidentd\fP will use the default port for the ident service ("auth" or
port 113). If the forwarded request fails, \fBoidentd\fP will fall back to
reading the \fB/etc/oidentd_masq.conf\fP file. In order for forwarding to work,
the machine to which the connection is forwarded must also be running
\fBoidentd\fP, and \fBoidentd\fP must be run with the \fB-P\fP switch
specifying the host that is forwarding the connections. If the ident daemon on
the host to which the connection is forwarded is capable of returning a fixed
string for any lookup (for example, the ident server built in to the mIRC
windows IRC client), it is not necessary to run \fBoidentd\fP on that host.
This option implies \fB--masquerade\fP.

.TP
.B "\-g or \-\-group=<group|GID>"
Run as the specified group or GID. If this option is not given, \fBoidentd\fP
falls back to running as \fIoidentd\fP, \fInobody\fP, \fInogroup\fP, or GID
\fI65534\fP, in this order. On systems where \fBoidentd\fP requires superuser
privileges, a warning is displayed and the group is not changed implicitly.

.TP
.B "\-i or \-\-foreground"
Run interactively, not as a daemon. This is useful for debugging, or when
running from a service manager such as daemontools.

.TP
.B "\-I or \-\-stdio"
Service only a single client request, then exit. The client is expected to
already be connected via stdin and stdout. This mode is useful when running
from listener utilities such as
.BR inetd (8),
.BR xinetd (8)
or
.BR tcpserver (8).
This option implies \fB-i\fP (run in foreground) also.

.TP
.B "\-l or \-\-limit=<number>"
Allow, at most, the specified number of open connections at once.

.TP
.B "\-m or \-\-masquerade"
Enable support for ident queries for masqueraded/NAT connections. See
.BR oidentd_masq.conf (5)
for details on configuring support for masqueraded/NAT connections.

.TP
.B "\-M or \-\-masquerade-first"
Check IP masquerading file before forwarding.

.TP
.B "\-o or \-\-other=[<string>]"
The string specified will be returned as the OS string by default for all
successful ident lookups. If no argument is given, "OTHER" will be returned
instead of the name of the operating system. The client side (with ident in
general, not just with \fBoidentd\fP) may interpret some requests as having
failed when some other string is returned instead of the name of the actual
operating system.

.TP
.B "\-p or \-\-port=<port>"
Listen on the specified port.

.TP
.B "\-P or \-\-proxy=<host>"
The specified host acts as a proxy, forwarding connections to us. This option
must be enabled when connections on the machine on which \fBoidentd\fP is
running are masqueraded through another host and the host through which the
connections are masqueraded forwards requests to us.

.TP
.B "\-q or \-\-quiet"
Quiet mode; do not log any status messages to syslog.

.TP
.B "\-S or \-\-nosyslog"
Log any status messages to stderr, not syslog. This is useful for debugging
or integration with external loggers such as
.BR multilog (8).

.TP
.B "\-t or \-\-timeout=<seconds>"
Sets the number of seconds to wait for input from a client before closing the
connection.

.TP
.B "\-u or \-\-user=<user|UID>"
Run as the specified user or UID. If this option is not given, \fBoidentd\fP
falls back to running as \fIoidentd\fP, \fInobody\fP, or UID \fI65534\fP, in
this order. On systems where \fBoidentd\fP requires superuser privileges, a
warning is displayed and the user is not changed implicitly.

.TP
.B "\-U or \-\-udb"
Perform lookups in the UDB shared memory tables, both for connections
originating on the local host and for masqueraded connections. When a match is
found, it will be used instead of the values supplied by the operating system,
for either masqueraded entries (with the \fB-m\fP flag) or normal TCP
connections. Entries in the table which don't match any local user will be
returned verbatim. This allows \fBoidentd\fP to cooperate with other programs
(e.g. RADIUS servers or proxies) to give valid replies for dynamic connections.

.TP
.B "\-r or \-\-reply=<string>"
Upon a failed lookup, the specified string will be returned to the client as if
the lookup had succeeded.

.TP
.B "\-v or \-\-version"
Display version information and exit.

.TP
.B "\-h or \-\-help"
Display options and exit.

.SH FILES
.TP
.B /etc/oidentd.conf
The system-wide configuration file.

.TP
.B /etc/oidentd_masq.conf
The NAT/IP masquerading mappings.

.TP
.B $HOME/.oidentd.conf
Per-user configuration file.

.SH AUTHOR
Janik Rabe <oidentd@janikrabe.com>
.br
https://oidentd.janikrabe.com
.PP
Originally written by Ryan McCabe <ryan@numb.org>.

.SH REPORTING BUGS
Please report any bugs to Janik Rabe <oidentd@janikrabe.com>.

.SH "SEE ALSO"
.BR oidentd.conf (5)
.BR oidentd_masq.conf (5)


NEWS
Unreleased changes


Changes in version 2.3.2

	* Option `--forward` now implies `--masquerade`.
	* Added a warning when the connection limit is exceeded.
	* Added systemd services and a socket file.
	* Added compile-time configuration to `--version` output.
	* Linux: fixed incorrect buffer lengths when compiled with `--disable-ipv6`.
	* Deprecated libudb (user database library) support.
	* Linux: fixed incorrect byte order in libnetfilter_conntrack queries.
	* Linux: fixed a null dereference with libnetfilter_conntrack on kernels
	  without a connection tracking file.
	* Build as a position-independent executable and with stack protection.
	* Fixed an invalid comment style in the default oidentd_masq.conf file.

Changes in version 2.3.1

	* Fixed build with libnetfilter_conntrack on Linux.

Changes in version 2.3.0

	* Added a forwarding capability.
	* Added Linux IPv6 masquerading support.
	* Added build information to `--version` output.
	* Always use libnetfilter_conntrack when available.
	* Fixed signed/unsigned bit shifting on Linux.
	* Minor bugfixes.

Changes in version 2.2.3

	* Fixed an out-of-buffer read/write issue.
	* Fixed several compiler warnings.
	* Fixed incorrect order of lport and fport in manual pages.
	* Fixed lookup of connections to IPv6-mapped IPv4 addresses.
	* Allow to, fport, from and lport to be specified in any order.
	* Warn when privileges cannot be dropped automatically.

Changes in version 2.2.2

	* Implicitly drop privileges only when running as root.
	* Try to run as oidentd:oidentd before falling back to nobody etc.
	* Search for configuration files in sysconfdir instead of `/etc`.
	* Deprecated option --forward-last in favor of --masquerade-first.

Changes in version 2.2.1

	* Fixed LIBNFCT_SUPPORT being defined incorrectly on some systems.

Changes in version 2.2.0

	* Security: privileges were not being dropped without the `--user`
	            and `--group` options.
	* Security: fixed a null pointer dereference with libudb support and
	            IP masquerading enabled.
	* Added a system-wide default configuration file.
	* Added a HACKING file with instructions for modifying oidentd.
	* Added a kernel support table (KERNEL_SUPPORT.md).
	* Added support for modern FreeBSD kernels.
	* Added support for modern OpenBSD kernels.
	* Added support for NetBSD 5+ (no IP masquerading support yet).
	* Added support for DragonFly BSD.
	* Added support for listening on multiple addresses.
	* Added flag `--forward-last`.
	* Bind to IPv6 by default.
	* Warn when `--debug` is used in a build without debugging support.
	* Warn when IP masquerading support is requested but unavailable.
	* Improved logging to make sure important errors are shown.
	* Bind socket before forking so that any errors are shown.
	* Follow RFC 1413 recommendations regarding white space in replies.
	* Modernized build system and replaced deprecated macros.
	* Updated all documentation and manual pages.
	* Fixed function inlined using a method incompatible with C99.
	* Fixed `--limit` option also enabling `--other`.
	* Fixed `--forward` option with a different destination port.
	* Fixed exit code when `--help` or `--version` was used.
	* Fixed configuration option `--disable-masq` having no effect.
	* Fixed various minor issues leading to compiler warnings.
	* Linux: handle local NAT properly instead of trying to forward.
	* Linux: fixed NAT when a different destination port was used.
	* Linux: fixed an issue with IPv4-mapped IPv6 addresses.
	* Linux: add masquerading support using `/proc/net/nf_conntrack`.
	* Linux: add masquerading support using `libnetfilter_conntrack`.
	* Linux: open connection tracking file only when necessary.
	* Removed Solaris 5.1+ support (build is broken due to kernel changes).

Changes in version 2.1.0

	* Fixed memory corruption on FreeBSD.
	* Changed maintainer to Janik Rabe <oidentd@janikrabe.com>.
	* Migrated project from CVS to git.

Changes in version 2.0.9

	* Internal changes.

Changes in version 2.0.8

	* So, I guess some things have changed in the past 3 years.
	* Fixed bugs.

Changes in version 2.0.7

	* Bug fixes.
	* Build fixes.

Changes in version 2.0.6

	* Bug fixes.
	* Build fixes for Darwin and FreeBSD.
	* Support for inetd/xinetd/daemontools.

Changes in version 2.0.5

	* Bugfixes for connection forwarding on BSD.
	* DNAT support on linux.
	* Configuration Parser/scanner cleanups.
	* IPv6 fixes.
	* FreeBSD 5 support.
	* Compile fixes.
	* Made connection forwarding work on recent versions of OpenBSD and on FreeBSD
	  when NAT is disabled.

Changes in version 2.0.4

	* Support for lookups using the netlink socket diagnostics interface on Linux.
	* Support for Solaris 9.
	* Support for NetBSD.
	* Support for Darwin.
	* Bugfixes for NAT on *BSD.

Changes in version 2.0.3

	* Fixed a bug in the configure script that caused a compile error on OpenBSD
	  3.0.

Changes in version 2.0.2

	* Added support for OpenBSD (>= 3.0) Packet Filter.
	* Allow multiple strings to be specified after a "reply" statement from which
	  one string will be selected at random.
	* A --limit command-line argument used to limit the number of connections that
	  can be open at any one time.
	* Fixed a bug that caused oidentd to chew up lots of CPU time on some
	  operating systems.
	* Note that oidentd only suports ipf NAT in FreeBSD (i.e. ipfw is not
	  supported).
	    - See the TODO file for slightly more info.

Changes in version 2.0.1

	* NAT support on OpenBSD 2.9 and greater has been enabled again.
	* Allow comments and whitespace in /etc/oidentd_masq.conf.
	* Fix a crash that occurred when oidentd reloaded its configuration file.

Changes in version 2.0.0

	* A new command-line flag, --config, can be used to specify the config file.
	* Support for OpenBSD 2.9
	* IP masquerading/NAT support for FreeBSD.
	* The bug that caused IP masquerading to fail was fixed.
	* More bug fixes.

Changes in version 1.9.9.1

	* Lots of bugfixes.
	* Support for Solaris 5.4 - 5.8


src/oidentd.c
/*
** oidentd.c - oidentd ident (RFC 1413) implementation.
** Copyright (c) 1998-2006 Ryan McCabe <ryan@numb.org>
** Copyright (c) 2018-2019 Janik Rabe  <oidentd@janikrabe.com>
**
** This program is free software; you can redistribute it and/or modify
** it under the terms of the GNU General Public License, version 2,
** as published by the Free Software Foundation.
**
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
**
** You should have received a copy of the GNU General Public License
** along with this program; if not, write to the Free Software
** Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
*/

#define _GNU_SOURCE

#include <config.h>

#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <signal.h>
#include <string.h>
#include <errno.h>
#include <pwd.h>
#include <syslog.h>
#include <pwd.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>

#include "oidentd.h"
#include "util.h"
#include "missing.h"
#include "inet_util.h"
#include "user_db.h"
#include "options.h"
#include "masq.h"

#if HAVE_LIBUDB
#	warning "libudb support is deprecated"
#	include <udb.h>
#endif

#ifndef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
static void sig_segv(int unused __notused) __noreturn;
static void sig_child(int sig);
static void sig_alarm(int unused __notused) __noreturn;
static void sig_hup(int unused);
#endif

static void copy_pw(const struct passwd *pw, struct passwd *pwd);
static void free_pw(struct passwd *pwd);

static void seed_prng(void);

static int service_request(int insock, int outsock);

u_int32_t timeout = DEFAULT_TIMEOUT;
u_int32_t connection_limit;
u_int32_t current_connections = 0;

uid_t target_uid;
gid_t target_gid;

char *ret_os;
char *failuser;
char *config_file;

in_port_t listen_port;
struct sockaddr_storage **addr;

int main(int argc, char **argv) {
	int *listen_fds = NULL;

	if (get_options(argc, argv) != 0)
		exit(EXIT_FAILURE);

	openlog("oidentd", LOG_PID | LOG_CONS | LOG_NDELAY, FACILITY);

	if (read_config(config_file) != 0) {
		o_log(LOG_CRIT, "Fatal: Error reading configuration file");
		exit(EXIT_FAILURE);
	}

	if (core_init() != 0) {
		if (opt_enabled(DEBUG_MSGS)) {
			o_log(LOG_CRIT, "Fatal: Error initializing core");
		} else {
			o_log(LOG_CRIT, "Fatal: Error initializing core (try --debug)");
		}
		exit(EXIT_FAILURE);
	}

	if (random_seed() != 0) {
		o_log(LOG_CRIT, "Fatal: Error seeding random number generator");
		exit(EXIT_FAILURE);
	}

	if (!opt_enabled(STDIO)) {
		listen_fds = setup_listen(addr, htons(listen_port));
		if (!listen_fds || listen_fds[0] == -1) {
			o_log(LOG_CRIT, "Fatal: Unable to set up listening socket");
			exit(EXIT_FAILURE);
		}
	}

	if (!opt_enabled(FOREGROUND) && go_background() == -1) {
		o_log(LOG_CRIT, "Fatal: Error creating daemon process");
		exit(EXIT_FAILURE);
	}

	if (k_open() != 0) {
		o_log(LOG_CRIT, "Fatal: Unable to open kmem device: %s", strerror(errno));
		exit(EXIT_FAILURE);
	}

#if LIBNFCT_SUPPORT
	if (!drop_privs_libnfct(target_uid, target_gid)) {
		o_log(LOG_CRIT, "Fatal: Failed to drop privileges (kernel)");
		exit(EXIT_FAILURE);
	}
#endif

	if (drop_privs(target_uid, target_gid) == -1) {
		o_log(LOG_CRIT, "Fatal: Failed to drop privileges (global)");
		exit(EXIT_FAILURE);
	}

#if HAVE_LIBUDB
	if (opt_enabled(USEUDB)) {
		if (udb_init(UDB_ENV_BASE_KEY) == 0) {
			o_log(LOG_CRIT, "Fatal: Can't open UDB shared memory tables");
			exit(EXIT_FAILURE);
		}
	}
#endif

#ifndef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
	signal(SIGALRM, sig_alarm);
	signal(SIGCHLD, sig_child);
	signal(SIGHUP, sig_hup);
	signal(SIGSEGV, sig_segv);
#endif

	if (opt_enabled(STDIO)) {
		service_request(fileno(stdin), fileno(stdout));
		exit(EXIT_SUCCESS);
	}

	for (;;) {
		fd_set rfds;
		int ret;
		size_t fdlen = 0;

		FD_ZERO(&rfds);

		do {
			int fd = listen_fds[fdlen++];
			FD_SET(fd, &rfds);
		} while (listen_fds[fdlen] != -1);

		ret = select(listen_fds[fdlen - 1] + 1, &rfds, NULL, NULL, NULL);
		if (ret > 0) {
			size_t i;

			for (i = 0; i < fdlen; ++i) {
				if (FD_ISSET(listen_fds[i], &rfds)) {
					int connectfd;

					connectfd = accept(listen_fds[i], NULL, NULL);
					if (connectfd == -1) {
						debug("accept: %s", strerror(errno));
						continue;
					}

					if (current_connections >= connection_limit) {
						o_log(NORMAL, "Connection limit exceeded; "
						              "closing incoming connection");
						close(connectfd);
						continue;
					}

					++current_connections;

					if (fork() == 0) {
						size_t idx;

						for (idx = 0; listen_fds[idx] != -1; ++idx)
							close(listen_fds[idx]);

						free(listen_fds);
						alarm(timeout);
						seed_prng();
						service_request(connectfd, connectfd);

						exit(EXIT_SUCCESS);
					}

					close(connectfd);
				}
			}
		}
	}

	exit(EXIT_SUCCESS);
}

/*
** Handle the client's request: read the client data and send the ident reply.
*/

static int service_request(int insock, int outsock) {
	int len;
	int ret;
	uid_t con_uid;
	int lport_temp;
	int fport_temp;
	in_port_t lport;
	in_port_t fport;
	char line[128];
	char suser[MAX_ULEN];
	char host_buf[MAX_HOSTLEN];
	char ip_buf[MAX_IPLEN];
	struct sockaddr_storage laddr, laddr6, faddr, faddr6;
	struct passwd *pw, pwd;

#ifdef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
	in_addr_t fuzz_faddr, fuzz_laddr;
	(void) inet_pton(AF_INET, "192.0.2.1", &fuzz_faddr);
	(void) inet_pton(AF_INET, "127.0.0.1", &fuzz_laddr);
	sin_setv4(fuzz_faddr, &faddr);
	sin_setv4(fuzz_laddr, &laddr);
	sin_set_port(49152, &faddr);
	sin_set_port(113, &faddr);
#else
	static socklen_t socklen = sizeof(struct sockaddr_storage);

	if (getpeername(insock, (struct sockaddr *) &faddr, &socklen) != 0) {
		debug("getpeername: %s", strerror(errno));
		return -1;
	}

	if (getsockname(insock, (struct sockaddr *) &laddr, &socklen) != 0) {
		debug("getsockname: %s", strerror(errno));
		return -1;
	}
#endif

	fport = htons(sin_port(&faddr));

#if WANT_IPV6
	laddr6 = laddr;
	faddr6 = faddr;

	if (laddr.ss_family == AF_INET6 &&
		IN6_IS_ADDR_V4MAPPED(&SIN6(&laddr)->sin6_addr))
	{
		struct in_addr in4;

		sin_extractv4(&SIN6(&laddr)->sin6_addr, &in4);
		sin_setv4(in4.s_addr, &laddr);

		sin_extractv4(&SIN6(&faddr)->sin6_addr, &in4);
		sin_setv4(in4.s_addr, &faddr);
	}
#endif

	get_ip(&faddr, ip_buf, sizeof(ip_buf));

	if (get_hostname(&faddr, host_buf, sizeof(host_buf)) != 0) {
		o_log(NORMAL, "Connection from %s:%d", ip_buf, fport);
		xstrncpy(host_buf, ip_buf, sizeof(host_buf));
	} else
		o_log(NORMAL, "Connection from %s (%s):%d", host_buf, ip_buf, fport);

	if (!sock_read(insock, line, sizeof(line)))
		return -1;

	len = sscanf(line, "%d , %d", &lport_temp, &fport_temp);
	if (len < 2) {
		debug("[%s] Malformed request: \"%s\"", host_buf, line);
		return 0;
	}

	if (!VALID_PORT(lport_temp) || !VALID_PORT(fport_temp)) {
		sockprintf(outsock, "%d,%d:ERROR:%s\r\n",
				lport_temp, fport_temp, ERROR("INVALID-PORT"));

		debug("[%s] %d , %d : ERROR : INVALID-PORT",
			host_buf, lport_temp, fport_temp);

		return 0;
	}

	lport = (in_port_t) lport_temp;
	fport = (in_port_t) fport_temp;

	/* User ID is unknown. */
	con_uid = MISSING_UID;

#if HAVE_LIBUDB
	if (opt_enabled(USEUDB)) {
		struct udb_lookup_res udb_res = get_udb_user(
				lport, fport, &laddr, &faddr, insock);
		if (udb_res.status == 2)
			return 0;
		con_uid = udb_res.uid;
	}
#endif

	if (con_uid == MISSING_UID && laddr.ss_family == AF_INET)
		con_uid = get_user4(htons(lport), htons(fport), &laddr, &faddr);

#if WANT_IPV6
	/*
	 * Check for IPv6-mapped IPv4 addresses. This ensures that the correct
	 * ident response is returned for connections to a mapped address.
	 */
	if (con_uid == MISSING_UID && laddr.ss_family == AF_INET) {
		struct sockaddr_storage laddr_m6, faddr_m6;
		struct in6_addr in6;

		sin_mapv4to6(&SIN4(&laddr)->sin_addr, &in6);
		sin_setv6(&in6, &laddr_m6);

		sin_mapv4to6(&SIN4(&faddr)->sin_addr, &in6);
		sin_setv6(&in6, &faddr_m6);

		con_uid = get_user6(htons(lport), htons(fport), &laddr_m6, &faddr_m6);
	}

	if (con_uid == MISSING_UID && laddr6.ss_family == AF_INET6)
		con_uid = get_user6(htons(lport), htons(fport), &laddr6, &faddr6);
#endif

	if (opt_enabled(MASQ)) {
		if (con_uid == MISSING_UID && laddr.ss_family == AF_INET)
			if (masq(insock, htons(lport), htons(fport), &laddr, &faddr) == 0)
				return 0;
	}

	if (con_uid == MISSING_UID) {
		if (failuser) {
			sockprintf(outsock, "%d,%d:USERID:%s:%s\r\n",
				lport, fport, ret_os, failuser);

			o_log(NORMAL, "[%s] Failed lookup: %d , %d : (returned %s)",
				host_buf, lport, fport, failuser);
		} else {
			sockprintf(outsock, "%d,%d:ERROR:%s\r\n",
				lport, fport, ERROR("NO-USER"));

			o_log(NORMAL, "[%s] %d , %d : ERROR : NO-USER",
				host_buf, lport, fport);
		}

		return 0;
	}

	pw = getpwuid(con_uid);
	if (!pw) {
		sockprintf(outsock, "%d,%d:ERROR:%s\r\n",
			lport, fport, ERROR("NO-USER"));

		debug("getpwuid(%d): %s", con_uid, strerror(errno));
		return 0;
	} else
		copy_pw(pw, &pwd);

	ret = get_ident(&pwd, lport, fport, &laddr, &faddr, suser, sizeof(suser));
	if (ret == -1) {
		sockprintf(outsock, "%d,%d:ERROR:%s\r\n",
			lport, fport, ERROR("HIDDEN-USER"));

		o_log(NORMAL, "[%s] %d , %d : HIDDEN-USER (%s)",
			host_buf, lport, fport, pwd.pw_name);

		goto out;
	}

	sockprintf(outsock, "%d,%d:USERID:%s:%s\r\n",
		lport, fport, ret_os, suser);

	o_log(NORMAL, "[%s] Successful lookup: %d , %d : %s (%s)",
		host_buf, lport, fport, pwd.pw_name, suser);

out:
	free_pw(&pwd);
	return 0;
}

/*
** Copy the needed fields from a passwd struct.
*/

static void copy_pw(const struct passwd *pw, struct passwd *pwd) {
	pwd->pw_name = xstrdup(pw->pw_name);
	pwd->pw_uid = pw->pw_uid;
	pwd->pw_gid = pw->pw_gid;
	pwd->pw_dir = xstrdup(pw->pw_dir);
}

/*
** Free a copied passwd struct.
*/

static void free_pw(struct passwd *pw) {
	free(pw->pw_name);
	free(pw->pw_dir);
}

#ifndef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
/*
** Handle SIGSEGV.
*/

static void sig_segv(int unused __notused) {
	o_log(LOG_CRIT, "Caught SIGSEGV; please report this to " PACKAGE_BUGREPORT);
	exit(EXIT_FAILURE);
}

/*
** Handle SIGCHLD.
*/

static void sig_child(int sig) {
	while (waitpid(-1, &sig, WNOHANG) > 0)
		--current_connections;

	signal(SIGCHLD, sig_child);
}

/*
** Handle SIGALRM.
*/

static void sig_alarm(int unused __notused) {
	o_log(NORMAL, "Timeout for request -- Closing connection");
	exit(EXIT_SUCCESS);
}

/*
** Handle SIGHUP - This causes oidentd to reload its configuration file.
*/

static void sig_hup(int unused __notused) {
	user_db_destroy();

	if (read_config(CONFFILE) != 0) {
		o_log(LOG_CRIT, "Error parsing configuration file");
		exit(EXIT_FAILURE);
	}
}
#endif

/*
** Seed the PRNG.
*/

static void seed_prng(void) {
#ifdef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
	srandom(0);
#else
	struct timeval tv;

	gettimeofday(&tv, NULL);
	srandom(tv.tv_sec ^ (tv.tv_usec << 11));
#endif
}
